src/middleware.ts

```diff
@@ -0,0 +1,118 @@
+  return locales.every(
+    (locale) =>
+      !pathname.startsWith(`/${locale}/`) &&
+      pathname !== `/${locale}` &&
+      // The root in default locale does not have a locale set
+      pathname !== '/'
+  );
+}
+
+type DefaultMatchesType = {
+  firstLocale?: string;
+  secondLocale?: string;
+};
+
+function getLocaleMatchesFromPathname(pathname: string): DefaultMatchesType {
+  const doubleLocalePrefix =
+    /^\/(?<firstLocale>fi|en|sv)\/(?<secondLocale>fi|en|sv)\//;
+  return (
+    pathname.match(doubleLocalePrefix)?.groups || {
+      firstLocale: undefined,
+      secondLocale: undefined,
+    }
+  );
+}
+
+/**
+ * Enforce prefix for default locale 'fi'
+ * https://github.com/vercel/next.js/discussions/18419
+ * @param req
+ */
+const prefixDefaultLocale = async (req: NextRequest) => {
+  const url = new URL(req.url);
+  const { pathname, search } = req.nextUrl;
+  const locale = getLocale(req);
+  // Check if there is any supported locale in the pathname
+  const nextUrlPathnameIsMissingLocale = isPathnameMissingLocale(pathname);
+  const requestUrlPathnameIsMissingLocale = isPathnameMissingLocale(
+    url.pathname
+  );
+
+  // The default locale needs to be redirected so that it uses the default language in URL.
+  if (req.nextUrl.locale === 'default') {
+    return NextResponse.redirect(
+      new URL(`/${DEFAULT_LANGUAGE}${pathname}${search}`, req.url)
+    );
+  }
+
+  // Redirect if there is no locale.
+  // Redirect is not needed for the default language
+  // If the current full request URL pathname is also missing the locale,
+  // set the default language as a locale.
+  // if (
+  //   nextUrlPathnameIsMissingLocale &&
+  //   requestUrlPathnameIsMissingLocale &&
+  //   locale !== DEFAULT_LANGUAGE
+  // ) {
+  //   // e.g. incoming request is /haku
+  //   // The new URL is now /fi/haku
+  //   return NextResponse.redirect(
+  //     new URL(`/${DEFAULT_LANGUAGE}${pathname}${search}`, req.url)
+  //   );
+  // }
+
+  // const { firstLocale, secondLocale } = getLocaleMatchesFromPathname(pathname);
+
+  // // Remove the double locale prefix which might occur
+  // // when CMS content has an internal link to another locale context.
+  // // NOTE: The locale will be automatically added as a prefix to
+  // // the next URL by the NextJS (under the hood).
+  // // The request's locale attribute should be used for that.
+  // if (firstLocale && secondLocale && firstLocale !== secondLocale) {
+  //   const nextPathname = pathname
+  //     .replace(new RegExp(`^/${firstLocale}`), '')
+  //     .replace(new RegExp(`^/${secondLocale}`), '');
+  //   const nextUrl = new URL(`${nextPathname}${search}`, req.url);
+  //   return NextResponse.redirect(nextUrl.href);
+  // }
+};
+
+export async function middleware(req: NextRequest) {
+  if (
+    requestType.isStaticFile(req) ||
+    requestType.isPagesFolderApi(req) ||
+    requestType.isPublicFile(req)
+  ) {
+    return;
+  }
+  return prefixDefaultLocale(req);
+}
```

Comment:
Some cleanup would be good to get rid of what is not needed or take into use what is still needed?

$ yarn lint
yarn run v1.22.21
$ eslint --ext js,ts,tsx src
palvelutarjotin-ui\src\middleware.ts
  44:10  warning  'getLocaleMatchesFromPathname' is defined but never used                @typescript-eslint/no-unused-vars
  63:9   warning  'locale' is assigned a value but never used                             @typescript-eslint/no-unused-vars
  65:9   warning  'nextUrlPathnameIsMissingLocale' is assigned a value but never used     @typescript-eslint/no-unused-vars
  66:9   warning  'requestUrlPathnameIsMissingLocale' is assigned a value but never used  @typescript-eslint/no-unused-vars
:heavy_multiplication_x: 4 problems (0 errors, 4 warnings)

Reply:
fixed